modules/core/src/test/java/org/apache/ignite/internal/processors/cache/NotAffinitySupplierWithMultipalRebalanceTest.java

```diff
@@ -0,0 +1,403 @@
+    public static final int BACKUPS = 2;
+
+    /** New nodes count. */
+    public static final int NEW_NODES = 3;
+
+    /** Cache with randezvous affinity. */
+    public static final String RENDEZVOUS_CACHE = DEFAULT_CACHE_NAME + "_rendezvous_aff";
+
+    /** Cache with custom affinity. */
+    public static final String CUSTOM_CACHE = DEFAULT_CACHE_NAME + "_specific_aff";
+
+    /** Persistent enabled. */
+    public boolean persistentEnabled;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setConsistentId(igniteInstanceName)
+            .setIncludeEventTypes(EVT_CACHE_REBALANCE_PART_MISSED)
+            .setCommunicationSpi(new TestRecordingCommunicationSpi())
+            .setDataStorageConfiguration(new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                    .setPersistenceEnabled(persistentEnabled)))
+            .setCacheConfiguration(
+                new CacheConfiguration(RENDEZVOUS_CACHE)
+                    .setBackups(BACKUPS),
+                new CacheConfiguration(CUSTOM_CACHE)
+                    .setBackups(BACKUPS)
+                    .setAffinity(new TestAffinity(4)));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+        cleanPersistenceDir();
+        System.clearProperty(IGNITE_PDS_WAL_REBALANCE_THRESHOLD);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testPersistentFullRebalance() throws Exception {
+        supplingFromOldBackup(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testInMemoryFullRebalance() throws Exception {
+        supplingFromOldBackup(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testPersistentHistoricalRebalance() throws Exception {
+        supplingHistoricalFromOldBackup(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void supplingHistoricalFromOldBackup(boolean persistentEnabled) throws Exception {
+        System.setProperty(IGNITE_PDS_WAL_REBALANCE_THRESHOLD, "0");
+
+        this.persistentEnabled = persistentEnabled;
+
+        IgniteEx ignite0 = startGrids(NODES_CNT);
+
+        for (int i = 1; i < 4; i++)
+            startGrid("new_" + i);
+
+        ignite0.cluster().active(true);
+
+        loadData(ignite0, RENDEZVOUS_CACHE);
+        loadData(ignite0, CUSTOM_CACHE);
+
+        ignite0.cluster().baselineAutoAdjustEnabled(false);
+
+        awaitPartitionMapExchange();
+
+        stopGrid("new_1");
+        stopGrid("new_2");
+
+        awaitPartitionMapExchange();
```

Comment:
do we really need this here ?

Reply:
I think need, because before PME completion we have instable topology.
Also for in-memory caches it can lead to re-balance.